src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -35,313 +40,370 @@ namespace xrt {
    * @param soft kernel CU index
    *
    */
-  skd::skd(xclDeviceHandle handle, int sk_meta_bohdl, int sk_bohdl, char *kname, uint32_t cu_index, unsigned char *uuid) {
-    strcpy(sk_name,kname);
-    parent_devHdl = handle;
-    cu_idx = cu_index;
-    sk_bo = sk_bohdl;
-    sk_meta_bo = sk_meta_bohdl;
-    memcpy(xclbin_uuid,uuid,sizeof(xclbin_uuid));
-
-    // Set sk_path according to sk_name
-    snprintf(sk_path, XRT_MAX_PATH_LENGTH, "%s%s%d", SOFT_KERNEL_FILE_PATH,
-	     sk_name,cu_idx);
+  skd::skd(const xclDeviceHandle handle, const int sk_meta_bohdl, const int sk_bohdl, const std::string kname, const uint32_t cu_index,
+	   unsigned char *uuid_in, const int parent_mem_bo_in, const uint64_t mem_start_paddr_in, const uint64_t mem_size_in)
+    : m_sk_name(kname), m_parent_devhdl(handle), m_cu_idx(cu_index), m_sk_bo(sk_bohdl), m_sk_meta_bo(sk_meta_bohdl),
+      m_xclbin_uuid(uuid_in), m_parent_bo_handle(parent_mem_bo_in), m_mem_start_paddr(mem_start_paddr_in), m_mem_size(mem_size_in),
+      m_sk_path(std::string(SOFT_KERNEL_FILE_PATH)+m_xclbin_uuid.to_string()+"/"+kname)
+  {
   }
 
   XCL_DRIVER_DLLESPEC
   int
   skd::init() {
-    void *buf = NULL;
-    xclBOProperties prop;
-    int ret = 0;
-
     // Create soft kernel file from sk_bo
-    if(createSoftKernelFile(parent_devHdl, sk_bo) != 0)
-      return -1;
-
-    /* Open and load the soft kernel. */
-    sk_handle = dlopen(sk_path, RTLD_LAZY | RTLD_GLOBAL);
-    char *errstr = dlerror();
-    if(errstr != NULL) {
-      syslog(LOG_ERR, "Dynamic Link error: %s\n", errstr);
-      return -1;
-    }
-    if (!sk_handle) {
-      syslog(LOG_ERR, "Cannot open %s\n", sk_path);
-      return -1;
+    int ret = create_softkernelfile(m_parent_devhdl, m_sk_bo);
```

Comment:
Since you wish any return values that are non-zero, the following code should work (haven't tested it):
```
int ret  = 0;
if ((ret = create_softkernelfile(m_parent_devhdl, m_sk_bo))
  return ret;
```